### PR TITLE
Add epiScanpy wrappers

### PR DESCRIPTION
**Pull request by gallardoalba**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/episcanpy/build_matrix.xml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    </macros>
+    <expand macro="bio_tools"/>
+    <expand macro="requirements"/>
+    <expand macro="version_command"/>
+    <command detect_errors="exit_code"><![CDATA[
+bgzip -c '$fragment_file' > fragments.gz &&
+tabix -p bed fragments.gz &&
+cat '$script_file' > '$hidden_output' &&
+python '$script_file' >> '$hidden_output' &&
+ls . >> '$hidden_output' &&
```
Comment (bgruening) [inline tools/episcanpy/build_matrix.xml:14]:
```suggestion
```

File: tools/episcanpy/macros.xml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        </citations>
+    </xml>
+    <xml name="version_command">
+        <version_command><![CDATA[python -c "import episcanpy as esc;print('episcanpy version: %s' % esc.__version__)"]]></version_command>
+    </xml>
+    <token name="@CMD@"><![CDATA[
+cp '$adata' 'anndata.h5ad' &&
+cat '$script_file' > '$hidden_output' &&
+python '$script_file' >> '$hidden_output' &&
+ls . >> '$hidden_output' &&
```
Comment (bgruening) [inline tools/episcanpy/macros.xml:28]:
```suggestion
```

File: tools/episcanpy/macros.xml [hunk 3]
```diff
@@ -0,0 +1,93 @@
+<macros>
+    <token name="@tool_version@">0.3.2</token>
+    <token name="@version_suffix@">0</token>
+    <token name="@profile@">21.01</token>
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="@tool_version@">episcanpy</requirement>
```
Comment (bgruening) [inline tools/episcanpy/macros.xml:7]: Please use those name for your variables: 

https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=version#tool-versions

File: tools/episcanpy/macros.xml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            <xref type="bio.tools">episcanpy</xref>
+        </xrefs>
+    </xml>
+    <xml name="citations">
+        <citations>
+            <citation type="doi">10.1038/s41467-021-25131-3</citation>
+        </citations>
+    </xml>
+    <xml name="version_command">
+        <version_command><![CDATA[python -c "import episcanpy as esc;print('episcanpy version: %s' % esc.__version__)"]]></version_command>
```
Comment (bgruening) [inline tools/episcanpy/macros.xml:22]:
```suggestion
        <version_command><![CDATA[python -c "import episcanpy as esc; print(f'esc.__version__')"]]></version_command>
```

File: tools/episcanpy/build_matrix.xml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    <tests>
+        <test>
+            <!-- count matrix -->
+            <param name="fragment_file" value="chrY.fragments.bed" />
+            <param name="peaks_file" value="chrY.peaks.bed" />
+            <conditional name="reference_chr">
+                <param name="chr_select" value="custom" />
+                <param name="chromosomes" value="Y" />
+            </conditional>
+            <output name="anndata_out" file="chrY.h5ad" ftype="h5ad" compare="sim_size"/>
```
Comment (bgruening) [inline tools/episcanpy/build_matrix.xml:70]: for hdf5 files we have some cool asserts: https://docs.galaxyproject.org/en/latest/dev/schema.html#has-h5-attribute

File: tools/episcanpy/macros.xml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        <data name="hidden_output" format="txt" label="Log file" hidden="true">
+            <filter>advanced_common['show_log']</filter>
+        </data>
+    </xml>
+    <token name="@CMD_anndata_write_outputs@"><![CDATA[
+adata.write('anndata.h5ad')
+with open('anndata_info.txt','w', encoding='utf-8') as ainfo:
+    print(adata, file=ainfo)
+]]></token>
+    <token name="@CMD_prettify_stdout@"><![CDATA[ | sed -r '1 s|AnnData object with (.+) = (.*)\s*|\1: \2|g' | sed "s|'||g"  | sed -r 's|^\s*(.*):\s(.*)|[\1]\n-    \2|g' | sed 's|, |\n-    |g'
```
Comment (bgruening) [inline tools/episcanpy/macros.xml:76]: Can you please add a comment what this is doing and why?

File: tools/episcanpy/build_matrix.xml (hunk 5 above)
Comment (bgruening) [inline tools/episcanpy/build_matrix.xml:70]: for hdf5 files we have some cool asserts: https://docs.galaxyproject.org/en/latest/dev/schema.html#has-h5-attribute

File: tools/episcanpy/preprocess.xml [hunk 7]
```diff
@@ -0,0 +1,393 @@
+<tool id="episcanpy_preprocess" name="Preprocessing" version="@tool_version@+galaxy@version_suffix@" profile="@profile@">
```
Comment (bgruening) [inline tools/episcanpy/preprocess.xml:1]: Preprocessing of what?

File: tools/episcanpy/preprocess.xml [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+                <option value="pp.coverage_cells">Coverage cells: Histogram of the number of open features (in the case of ATAC-seq data) per cell, using 'pp.coverage_cells'</option>
+                <option value="pp.coverage_features">Coverage features: Distribution of the feature commoness in cells, using 'pp.coverage_features'</option>
+                <option value="pp.select_var_feature">Select the most variable features, 'pp.select_var_feature'</option>
+                <option value="pp.cal_var">Show distribution plots of cells sharing features and variability score 'pp.cal_var'</option>
+                <option value="pp.variability_features">Computes variability score to rank the most variable features across all cells, using 'pp.variability_features'</option>
+            </param>
+            <when value="pp.filter_cells">
+                <conditional name="filter">
+                    <param argument="filter" type="select" label="Filter" help="Filter mode">
+                        <option value="min_counts">Minimum number of counts</option>
```
Comment (bgruening) [inline tools/episcanpy/preprocess.xml:120]: You can only filter for one of them at a time?

File: tools/episcanpy/preprocess.xml [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+        <expand macro="inputs_common_advanced"/>
+    </inputs>
+    <outputs>
+        <expand macro="anndata_outputs"/>
+        <data name="out_png" format="png" from_work_dir="plot.png" label="${tool.name} (${method.method}) on ${on_string}">
+            <filter>method['method'] != 'pp.filter_cells' and method['method'] != 'pp.filter_features' and method['method'] != 'nb_feat_log' and method['method'] != 'select_var_feature'</filter>
+        </data>
+    </outputs>
+    <tests>
+        <test>
```
Comment (bgruening) [inline tools/episcanpy/preprocess.xml:193]: you can count the output files here with `expect_num_outputs`

File: tools/episcanpy/cluster_embed.xml [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+<!--                <param name="find_genes_feature_type" value="transcript"/>-->
+<!--                <param name="find_genes_annotation" value="HAVANA"/>-->
+<!--                <param name="find_genes_raw" value="False"/>-->
+<!--            </conditional>-->
+<!--            <section name="advanced_common">-->
+<!--                <param name="show_log" value="true" />-->
+<!--            </section>-->
+<!--            <output name="anndata_out" file="anndata.h5ad" ftype="h5ad"/>-->
+<!--        </test>-->
+        <test>
```
Comment (bgruening) [inline tools/episcanpy/cluster_embed.xml:274]: add expect_num_outputs

File: tools/episcanpy/preprocess.xml (hunk 8 above)
Comment (pavanvidem) [reply tools/episcanpy/preprocess.xml:120]: Yes, only one at a time. Filtering is done sequentially. The function accepts only one of the filtering criteria at a time.

File: tools/episcanpy/macros.xml (hunk 3 above)
Comment (pavanvidem) [reply tools/episcanpy/macros.xml:7]: is there something missing here?

Comment (bgruening) [reply tools/episcanpy/macros.xml:7]: @TOOL_VERSION@ capital letters, please. Our bot is expecting those.

**Comment by pavanvidem** [comment]
I couldn't figure out the problem with the build_matrix test. It is successful on my computer with the `--biocontainers` option. Is there a way to look at the error message?

**Comment by pavanvidem** [comment]
It was tabix dependency. It is ready now.

File: tools/episcanpy/build_matrix.xml [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    chrY	2650957	2650995	GACTAACAGCAACGGT-1	1
+    chrY	2651205	2651265	TCAAAGCTCAAAGTAG-1	1
+    chrY	2651215	2651268	TCACAAGGTCAAGACG-1	1
+
+- Features file. A plain BED file with peak locations or narrowPeak file from MACS2.
+
+**Output**
+
+- Count matrix in Anndata format.
+    </help> -->
```
Comment (bgruening) [inline tools/episcanpy/build_matrix.xml:131]:
```suggestion
```

File: tools/episcanpy/macros.xml [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    ]]></token>
+    <xml name="coverage_params">
+        <param argument="binary" type="boolean" truevalue="True" falsevalue="False" checked="false" 
+            label="Binary" help="To reduce computing time, you can specify if the matrix is already binary" />
```
Comment (bgruening) [inline tools/episcanpy/macros.xml:84]: Is this parameter really needed? Should this not be decoded in the datatype? mtx.gz vs. mtx or similar?
If we really need this param the label should be better explained.

File: tools/episcanpy/macros.xml [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        <param argument="log" type="boolean" truevalue="True" falsevalue="False" checked="false"
+            label="Log transform?" />
+        <param argument="threshold" type="float" min="0" value="" label="Threshold"/>
```
Comment (bgruening) [inline tools/episcanpy/macros.xml:87]: Threshold of what?

File: tools/episcanpy/macros.xml [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        <param argument="bins" type="integer" min="0" value="50" label="Bins"/>
```
Comment (bgruening) [inline tools/episcanpy/macros.xml:88]: Bins of what?

File: tools/episcanpy/cluster_embed.xml [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+                    <option value="minkowski">minkowski</option>
+                    <option value="rogerstanimoto">rogerstanimoto</option>
+                    <option value="russelrao">russelrao</option>
+                    <option value="seuclidean">seuclidean</option>
+                    <option value="sokalmichener">sokalmichener</option>
+                    <option value="sokalsneath">sokalsneath</option>
+                    <option value="sqeuclidean">sqeuclidean</option>
+                    <option value="yule">yule</option>
+                </param>
+                <param name="lazy_min_dist" value="0.5" type="float" label="The effective minimum distance between embedded points" help="(min_dist)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:139]:
```suggestion
                <param name="lazy_min_dist" value="0.5" min="0" type="float" label="The effective minimum distance between embedded points" help="(min_dist)"/>
```

File: tools/episcanpy/cluster_embed.xml [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+                    <option value="True" selected="true">True</option>
+                    <option value="False">False</option>
+                </param>
+                <param name="lazy_svd_solver" type="select" label="SVD solver to use" help="(svd_solver)">
+                    <option value="arpack" selected="true">arpack (for the ARPACK wrapper in SciPy)</option>
+                    <option value="randomized">randomized (for the randomized algorithm due to Halko (2009)</option>
+                    <option value="auto">auto (chooses automatically depending on the size of the problem)</option>
+                    <option value="lobpcg">lobpcg (an alternative SciPy solver)</option>
+                </param>
+                <param name="lazy_nb_pcs" value="50" type="integer" label="Number of principal component computed for PCA (and therefore neighbors, tsne and umap)" help="(nb_pcs)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:107]:
```suggestion
                <param name="lazy_nb_pcs" value="50" min="'0" type="integer" label="Number of principal component computed for PCA (and therefore neighbors, tsne and umap)" help="(nb_pcs)"/>
```

File: tools/episcanpy/cluster_embed.xml [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="lazy_n_neighbors" value="15" type="integer" label="Size of the local neighborhood (number of neighboring data points) used for manifold approximation" help="(n_neighbors)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:108]:
```suggestion
                <param name="lazy_n_neighbors" value="15" min="0" type="integer" label="Size of the local neighborhood (number of neighboring data points) used for manifold approximation" help="(n_neighbors)"/>
```

File: tools/episcanpy/cluster_embed.xml [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="lazy_perplexity" value="30" type="integer" label="Perplexity (number of nearest neighbors used in other manifold learning algorithms)" help="(perplexity)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:109]:
```suggestion
                <param name="lazy_perplexity" value="30" min="0" type="integer" label="Perplexity (number of nearest neighbors used in other manifold learning algorithms)" help="(perplexity)"/>
```

File: tools/episcanpy/cluster_embed.xml [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+                    <option value="sqeuclidean">sqeuclidean</option>
+                    <option value="yule">yule</option>
+                </param>
+                <param name="lazy_min_dist" value="0.5" type="float" label="The effective minimum distance between embedded points" help="(min_dist)"/>
+                <param name="lazy_spread" value="1.0" type="float" label="The effective scale of embedded points" help="(spread)"/>
+                <param name="lazy_use_highly_variable" type="select" label="Use highly variable genes only" help="(use_highly_variable)">
+                    <option value="True">True</option>
+                    <option value="False" selected="true">False</option>
+                </param>
+                <param name="lazy_n_components" value="2" type="integer" label="The number of dimensions of the UMAP embedding" help="(n_components)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:145]:
```suggestion
                <param name="lazy_n_components" value="2" min="0" type="integer" label="The number of dimensions of the UMAP embedding" help="(n_components)"/>
```

File: tools/episcanpy/cluster_embed.xml [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+            </when>
+            <when value="tl.rank_features">
+                <param name="rank_features_groupby" value="louvain" type="text" label="The key of the observations grouping to consider" help="(groupby)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:148]:
```suggestion
                <param name="rank_features_groupby" value="louvain" type="text" label="The key of the observations grouping to consider" help="(groupby)">
                    <sanitizer invalid_char="">
                        <valid initial="string.letters,string.digits">
                            <add value="_" />
                            <add value="-" />
                            <add value=" " />
                            <add value="." />
                        </valid>
                    </sanitizer>
                    <validator type="regex">[0-9a-zA-Z_. -]+</validator>
                </param>
```

File: tools/episcanpy/cluster_embed.xml [hunk 21]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="rank_features_use_raw" type="select" label="Use raw attribute of Anndata if present" help="(use_raw)">
+                    <option value="True" selected="true">True</option>
+                    <option value="False">False</option>
+                </param>
+                <param name="rank_features_groups" value="all" type="text" label="Subset of groups, to which comparison shall be restricted" help="(groups)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:153]:
```suggestion
                <param name="rank_features_groups" value="all" type="text" label="Subset of groups, to which comparison shall be restricted" help="(groups)">
                    <sanitizer invalid_char="">
                        <valid initial="string.letters,string.digits">
                            <add value="_" />
                            <add value="-" />
                            <add value="." />
                            <add value=" " />
                            <add value="," />
                        </valid>
                    </sanitizer>
                    <validator type="regex">[0-9a-zA-Z_., -]+</validator>
                </param>
```

File: tools/episcanpy/macros.xml (hunk 12 above)
Comment (pavanvidem) [reply tools/episcanpy/macros.xml]: Unfortunately, it cannot automatically decode it. It is a function parameter indicating whether the input matrix is already binarized or not.

File: tools/episcanpy/cluster_embed.xml [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="rank_features_reference" value="rest" type="text" label="Compare each group with respect to this group" help="(reference)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:154]:
```suggestion
                <param name="rank_features_reference" value="rest" type="text" label="Compare each group with respect to this group" help="(reference)">
                    <sanitizer invalid_char="">
                        <valid initial="string.letters,string.digits">
                            <add value="_" />
                            <add value="-" />
                            <add value="." />
                            <add value=" " />
                            <add value="," />
                        </valid>
                    </sanitizer>
                    <validator type="regex">[0-9a-zA-Z_., -]+</validator>
                </param>
```

File: tools/episcanpy/cluster_embed.xml [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="rank_features_n_features" value="100" type="integer" label="The number of features that appear in the returned tables" help="(n_features)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:155]:
```suggestion
                <param name="rank_features_n_features" value="100" min="1" type="integer" label="The number of features that appear in the returned tables" help="(n_features)"/>
```

File: tools/episcanpy/cluster_embed.xml [hunk 24]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="rank_features_rankby_abs" type="select" label="Rank genes by the absolute value of the score, not by the score" help="(rankby_abs)">
+                    <option value="True" >True</option>
+                    <option value="False" selected="true">False</option>
+                </param>
+                <param name="rank_features_key_added" value="rank_features_groups" type="text" label="The key in adata.uns information is saved to" help="(key_added)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:160]:
```suggestion
                <param name="rank_features_key_added" value="rank_features_groups" type="text" label="The key in adata.uns information is saved to" help="(key_added)">
                    <sanitizer invalid_char="">
                        <valid initial="string.letters,string.digits">
                            <add value="_" />
                            <add value="-" />
                            <add value="." />
                            <add value=" " />
                            <add value="," />
                        </valid>
                    </sanitizer>
                    <validator type="regex">[0-9a-zA-Z_., -]+</validator>
                </param>
```

File: tools/episcanpy/cluster_embed.xml [hunk 25]
```diff
@@ -0,0 +1,363 @@
...
+                    <option value="t-test_overestim_var">t-test_overestim_var</option>
+                    <option value="wilcoxon">Wilcoxon rank sum</option>
+                </param>
+                <param name="rank_features_corr_method" value="benjamini-hochberg" type="select" label="p-value correction method" help="(corr_method)">
+                    <option value="benjamini-hochberg">Benjamini Hochberg</option>
+                    <option value="bonferroni">Bonferroni</option>
+                </param>
+            </when>
+            <when value="tl.find_genes">
+                <param name="find_genes_gtf_file" type="data" format="gtf" label="Annotation GTF file" help="(gtf_file)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:174]: Does it also accept GFF files? I tried to find information in the docs, but there's nothing about that.

File: tools/episcanpy/cluster_embed.xml [hunk 26]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="find_genes_key_added" value="transcript_annotation" type="text" label="Key added" help="(key_added)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:175]:
```suggestion
                <param name="find_genes_key_added" value="transcript_annotation" type="text" label="Key added" help="(key_added)">
                    <sanitizer invalid_char="">
                        <valid initial="string.letters,string.digits">
                            <add value="_" />
                            <add value="-" />
                            <add value="." />
                            <add value=" " />
                            <add value="," />
                        </valid>
                    </sanitizer>
                    <validator type="regex">[0-9a-zA-Z_., -]+</validator>
                </param>
```

File: tools/episcanpy/cluster_embed.xml [hunk 27]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="find_genes_upstream" value="2000" type="integer" label="Upstream" help="(upstream)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:176]:
```suggestion
                <param name="find_genes_upstream" value="2000" min="0" type="integer" label="Upstream" help="(upstream)"/>
```

File: tools/episcanpy/cluster_embed.xml [hunk 28]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="find_genes_feature_type" value="transcript" type="text" label="Feature type" help="(feature_type)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:177]:
```suggestion
                <param name="find_genes_feature_type" value="transcript" type="text" label="Feature type" help="(feature_type)">
                    <sanitizer invalid_char="">
                        <valid initial="string.letters,string.digits">
                            <add value="_" />
                            <add value="-" />
                            <add value="." />
                            <add value=" " />
                            <add value="," />
                        </valid>
                    </sanitizer>
                    <validator type="regex">[0-9a-zA-Z_., -]+</validator>
                </param>
```

File: tools/episcanpy/cluster_embed.xml [hunk 29]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="find_genes_annotation" value="HAVANA" type="text" label="Annotation" help="(annotation)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:178]:
```suggestion
                <param name="find_genes_annotation" value="HAVANA" type="text" label="Annotation" help="(annotation)">
                    <sanitizer invalid_char="">
                        <valid initial="string.letters,string.digits">
                            <add value="_" />
                            <add value="-" />
                            <add value="." />
                            <add value=" " />
                            <add value="," />
                        </valid>
                    </sanitizer>
                    <validator type="regex">[0-9a-zA-Z_., -]+</validator>
                </param>
```

File: tools/episcanpy/cluster_embed.xml [hunk 30]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="find_genes_raw" type="select" label="Raw?" help="(raw)">
+                    <option value="True">True</option>
+                    <option value="False" selected="true">False</option>
+                </param>
+            </when>
+            <when value="tl.get_n_clusters">
+                <param name="get_n_clusters_n_cluster" value="14" type="integer" label="Number of clusters" help="(n_cluster)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:185]:
```suggestion
                <param name="get_n_clusters_n_cluster" value="14" min="1" type="integer" label="Number of clusters" help="(n_cluster)"/>
```

File: tools/episcanpy/cluster_embed.xml [hunk 31]
```diff
@@ -0,0 +1,363 @@
...
+                <param name="get_n_clusters_method" type="select" label="Clustering method to use" help="(method)">
+                    <option value="leiden" selected="true">leiden</option>
+                    <option value="louvain">louvain</option>
+                </param>
+                <param name="get_n_clusters_range_min" value="0" type="integer" label="Range minimum" help="(range_min)"/>
+                <param name="get_n_clusters_range_max" value="3" type="integer" label="Range maximum" help="(range_max)"/>
+                <param name="get_n_clusters_max_steps" value="20" type="integer" label="Maximum number of steps" help="(max_steps)"/>
+                <param name="get_n_clusters_key_added" value="None" type="text" label="Variable name in obs" help="(key_added)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:193]:
```suggestion
                <param name="get_n_clusters_range_min" value="0" min="0" type="integer" label="Range minimum" help="(range_min)"/>
                <param name="get_n_clusters_range_max" value="3" min="1" type="integer" label="Range maximum" help="(range_max)"/>
                <param name="get_n_clusters_max_steps" value="20" min="1" type="integer" label="Maximum number of steps" help="(max_steps)"/>
                <param name="get_n_clusters_key_added" value="None" type="text" label="Variable name in obs" help="(key_added)">
                    <sanitizer invalid_char="">
                        <valid initial="string.letters,string.digits">
                            <add value="_" />
                            <add value="-" />
                            <add value="." />
                            <add value=" " />
                            <add value="," />
                        </valid>
                    </sanitizer>
                    <validator type="regex">[0-9a-zA-Z_., -]+</validator>
                </param>
```

File: tools/episcanpy/cluster_embed.xml [hunk 32]
```diff
@@ -0,0 +1,363 @@
...
+            </when>
+            <when value="tl.kmeans">
+                <param name="kmeans_num_clusters" value="14" type="integer" label="Number of clusters" help="(num_clusters)"/>
+            </when>
+            <when value="tl.hc">
+                <param name="hc_num_clusters" value="14" type="integer" label="Number of clusters" help="(num_clusters)"/>
```
Comment (gallardoalba) [inline tools/episcanpy/cluster_embed.xml:199]:
```suggestion
                <param name="kmeans_num_clusters" value="14" min="1" type="integer" label="Number of clusters" help="(num_clusters)"/>
            </when>
            <when value="tl.hc">
                <param name="hc_num_clusters" value="14" min="1" type="integer" label="Number of clusters" help="(num_clusters)"/>
```

**Review by bgruening — APPROVED** [review]

**Comment by pavanvidem** [comment]
thanks a lot @bgruening @gallardoalba. Is it ready to merge?

**Comment by gallardoalba** [comment]
> thanks a lot @bgruening @gallardoalba. Is it ready to merge?

Yes, I think so.

**Pull request merged**